core/drivers/scmi-msg/voltage_domain.c

```diff
@@ -0,0 +1,415 @@
+
+#define LEVELS_BY_STEP \
+	SCMI_VOLTAGE_DOMAIN_LEVELS_FLAGS(3, SCMI_VOLTD_LEVELS_FORMAT_RANGE, 0)
+
+static void scmi_voltd_describe_levels(struct scmi_msg *msg)
+{
+	const struct scmi_voltd_describe_levels_a2p *in_args = (void *)msg->in;
+	size_t nb_levels = 0;
+	unsigned int domain_id = 0;
+	int32_t status = SCMI_GENERIC_ERROR;
+
+	if (msg->in_size != sizeof(*in_args)) {
+		status = SCMI_PROTOCOL_ERROR;
+		goto out;
+	}
+
+	if (in_args->domain_id >= plat_scmi_voltd_count(msg->agent_id)) {
+		status = SCMI_INVALID_PARAMETERS;
+		goto out;
+	}
+
+	domain_id = confine_array_index(in_args->domain_id,
+					plat_scmi_voltd_count(msg->agent_id));
+
+	/* Platform may support array rate description */
+	status = plat_scmi_voltd_levels_array(msg->agent_id, domain_id, 0, NULL,
+					      &nb_levels);
+	if (status == SCMI_SUCCESS) {
+		/* Use the stack to get the returned portion of levels array */
+		long plat_levels[LEVELS_ARRAY_SIZE_MAX];
+		size_t ret_nb = 0;
+		size_t rem_nb = 0;
+
+		if (in_args->level_index >= nb_levels) {
+			status = SCMI_INVALID_PARAMETERS;
+			goto out;
+		}
+
+		ret_nb = MIN(ARRAY_SIZE(plat_levels),
+			     nb_levels - in_args->level_index);
+		rem_nb = nb_levels - in_args->level_index - ret_nb;
+
+		status =  plat_scmi_voltd_levels_array(msg->agent_id, domain_id,
+						       in_args->level_index,
+						       plat_levels,
+						       &ret_nb);
+		if (status == SCMI_SUCCESS) {
+			struct scmi_voltd_describe_levels_p2a out_args = {
+				.flags = LEVELS_BY_ARRAY(ret_nb, rem_nb),
+				.status = SCMI_SUCCESS,
+			};
+			uint32_t *voltage = NULL;
+			size_t i = 0;
+
+			/* By construction these values are 32bit aligned */
+			voltage = (uint32_t *)(uintptr_t)(msg->out +
+							  sizeof(out_args));
+
+			for (i = 0; i < ret_nb; i++)
+				voltage[i] = plat_levels[i];
+
+			memcpy(msg->out, &out_args, sizeof(out_args));
+
+			msg->out_size_out = sizeof(out_args) +
+					    ret_nb * sizeof(uint32_t);
+		}
+	} else if (status == SCMI_NOT_SUPPORTED) {
+		long triplet[3] = { 0, 0, 0 };
+
+		/* Platform may support min/max/step triplet description */
+		status =  plat_scmi_voltd_levels_by_step(msg->agent_id,
+							 domain_id, triplet);
+		if (status == SCMI_SUCCESS) {
+			struct scmi_voltd_describe_levels_p2a out_args = {
+				.flags = LEVELS_BY_STEP,
+				.status = SCMI_SUCCESS,
+			};
+			int32_t *voltage = NULL;
+
+			/* By construction these values are 32bit aligned */
+			voltage = (int32_t *)(uintptr_t)(msg->out +
+							 sizeof(out_args));
+			voltage[0] = triplet[0];
+			voltage[1] = triplet[1];
+			voltage[2] = triplet[2];
+
+			memcpy(msg->out, &out_args, sizeof(out_args));
+
+			msg->out_size_out = sizeof(out_args) +
+					    3 * sizeof(uint32_t);
+		}
+	} else {
+		/* Fallthrough generic exit sequence below with error status */
+	}
+
+out:
+	if (status) {
+		scmi_status_response(msg, status);
+	} else {
+		/*
+		 * Message payload is already written to msg->out, and
+		 * msg->out_size_out updated.
+		 */
+	}
```

Comment:
Minor: I don't think the empty `else { ... }` blocks are really helpful, and the first one above causes a checkpatch warning. Up to you to keep them or not ;-)

Reply:
Ok, i removed them in fixup commit.